kyo-prelude/shared/src/main/scala/kyo/Poll.scala

```diff
@@ -0,0 +1,281 @@
+  *   - Once Absent is received, the consumer should stop polling as the stream has terminated
+  *   - Backpressure is maintained through the Ack responses:
+  *     - Continue signals readiness to receive more values
+  *     - Stop indicates the consumer wants to pause receiving values
+  *
+  * The effect enables building streaming data pipelines with controlled consumption rates. Handlers can process values at their own pace by
+  * returning appropriate Ack responses, while respecting stream termination signals.
+  *
+  * @tparam V
+  *   The type of values being polled from the data source.
+  */
+sealed trait Poll[V] extends ArrowEffect[Const[Ack], Const[Maybe[V]]]
+
+object Poll:
+
+    /** Attempts to poll a single value with acknowledgment.
+      *
+      * Requests a value from the data source while providing an acknowledgment to control flow. The acknowledgment signals the source about
+      * consumer readiness and backpressure.
+      *
+      * @param ack
+      *   The acknowledgment controlling value emission
+      * @return
+      *   Maybe containing the polled value if available, or Absent if stream ended
+      */
+    inline def one[V](inline ack: Ack)(
+        using
+        inline tag: Tag[Poll[V]],
+        inline frame: Frame
+    ): Maybe[V] < Poll[V] =
+        ArrowEffect.suspend[Unit](tag, ack)
+
+    /** Attempts to poll a single value with Continue acknowledgement.
+      *
+      * @return
+      *   A computation that produces Maybe containing the polled value if available
+      */
+    inline def one[V](
+        using
+        inline tag: Tag[Poll[V]],
+        inline frame: Frame
+    ): Maybe[V] < Poll[V] =
+        one(Ack.Continue())
+
+    final class ApplyOps[V](dummy: Unit) extends AnyVal:
+
+        /** Processes polled values with a function. Values are processed until n is reached or the stream completes.
+          *
+          * @param n
+          *   Maximum number of values to process
+          * @param f
+          *   Function to apply to each value
+          * @return
+          *   A computation that processes values until completion or limit reached
+          */
+        def apply[S](n: Int)(f: V => Unit < S)(using tag: Tag[Poll[V]], frame: Frame): Unit < (Poll[V] & S) =
+            Loop.indexed { idx =>
+                if idx == n then Loop.done
+                else
+                    Poll.andMap[V] {
+                        case Present(v) => f(v).map(Loop.continue)
+                        case Absent     => Loop.done
+                    }
+            }
+
+        /** Processes polled values with a function until the stream completes.
+          *
+          * @param f
+          *   Function to apply to each value
+          * @return
+          *   A computation that processes values until completion
+          */
+        def apply[S](f: V => Unit < S)(using tag: Tag[Poll[V]], frame: Frame): Unit < (Poll[V] & S) =
+            Loop(()) { _ =>
```

Comment:
Should we add a `apply0`? Followup is fine.

Reply:
I tried some time ago but had inference issues